File: 78-subsets/subsets.cpp
class Solution {
public:
    vector<vector<int>>ans;
    void solve(int i,int n,vector<int>empty,vector<int>&nums){
        if(i==n){
            ans.push_back(empty);
            return;
        }
        empty.push_back(nums[i]);
        solve(i+1,n,empty,nums);
        empty.pop_back();
        solve(i+1,n,empty,nums);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        int n=nums.size();
        vector<int>empty;
        solve(0,n,empty,nums);
        return ans;
        
    }
};

File: 78-subsets/subsets.java
class Solution {
    List<List<Integer>>result = new ArrayList<>();
    ArrayList<Integer> empty = new ArrayList<>();
    public void solve(int i,int[]nums){
        if(i==nums.length){
            result.add(new ArrayList<>(empty));
            return;
        }
        empty.add(nums[i]);
        solve(i+1,nums);
        empty.remove(empty.size()-1);
        solve(i+1,nums);
    }
    public List<List<Integer>> subsets(int[] nums) {
        solve(0,nums);
        return result;
    }
}